File: code/packages/fsharp/vhdl-parser/tests/CodingAdventures.VhdlParser.Tests/VhdlParserTests.fs
namespace CodingAdventures.VhdlParser.Tests

open CodingAdventures.VhdlParser.FSharp
open Xunit

module VhdlParserTests =
    [<Fact>]
    let parsesSample () =
        let ast = VhdlParser.ParseVhdl("entity top is end entity top;")

        Assert.Equal("design_file", ast.RuleName)
        Assert.True(ast.DescendantCount() > 0)

    [<Fact>]
    let defaultVersionMatchesExplicitVersion () =
        Assert.Equal(
            VhdlParser.ParseVhdl("entity top is end entity top;").RuleName,
            VhdlParser.ParseVhdl("entity top is end entity top;", VhdlParser.DefaultVersion).RuleName)

    [<Fact>]
    let rejectsUnknownVersion () =
        let error = Assert.Throws<System.ArgumentException>(fun () -> VhdlParser.ParseVhdl("entity top is end entity top;", "2099") |> ignore)
        Assert.Contains("Unknown VHDL version", error.Message)

File: code/packages/csharp/vhdl-parser/tests/CodingAdventures.VhdlParser.Tests/VhdlParserTests.cs
namespace CodingAdventures.VhdlParser.Tests;

public sealed class VhdlParserTests
{
    [Fact]
    public void ParsesSample()
    {
        var ast = CodingAdventures.VhdlParser.VhdlParser.ParseVhdl("entity top is end entity top;");

        Assert.Equal("design_file", ast.RuleName);
        Assert.True(ast.DescendantCount() > 0);
    }

    [Fact]
    public void DefaultVersionMatchesExplicitVersion()
    {
        Assert.Equal(
            CodingAdventures.VhdlParser.VhdlParser.ParseVhdl("entity top is end entity top;").RuleName,
            CodingAdventures.VhdlParser.VhdlParser.ParseVhdl("entity top is end entity top;", CodingAdventures.VhdlParser.VhdlParser.DefaultVersion).RuleName);
    }

    [Fact]
    public void RejectsUnknownVersion()
    {
        var error = Assert.Throws<ArgumentException>(() => CodingAdventures.VhdlParser.VhdlParser.ParseVhdl("entity top is end entity top;", "2099"));
        Assert.Contains("Unknown VHDL version", error.Message);
    }
}
